hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineImpl.java

```diff
@@ -317,30 +335,61 @@ private synchronized void addPendingData(Data data) {
         pendingData.add(data);
     }
 
+    private synchronized void addPendingEvent(Collection<Event> event) {
+        pendingEvent.addAll(event);
+    }
+
+    private synchronized void addPendingEvent(Event event) {
+        pendingEvent.add(event);
+    }
+
     private synchronized Collection<Data> getAndClearPendingData() {
         Collection<Data> result = new ArrayList<>(pendingData);
         pendingData.clear();
         return result;
     }
 
+    private synchronized Collection<Event> getAndClearPendingEvent() {
+        Collection<Event> result = new ArrayList<>(pendingEvent);
+        pendingEvent.clear();
+        return result;
+    }
+
     private class RulesInvoker extends TimerTask {
         @Override
         public void run() {
             int numTimeouts = checkPendingTimeouts();
 
-            if (!pendingData.isEmpty() || numTimeouts > 0) {
+            if (!pendingData.isEmpty() || !pendingEvent.isEmpty() || numTimeouts > 0) {
                 Collection<Data> newData = getAndClearPendingData();
+                Collection<Event> newEvent = getAndClearPendingEvent();
+
+                if (!newEvent.isEmpty()) {
+                    /*
+                        Incoming Events are added to the result events list.
+                        Events are all persisted at the end of the inference.
+                        With this we will have on events list:
+                        - Input events.
+                        - New events generates as result of the inference process.
+                     */
+                    events.addAll(newEvent);
+                }
```

Comment:
I'm not sure I understand this.  Aren't API-inserted Events persisted independently of being sent into the engine?  I don't think we should be worrying about persisting API-inserted event here in this code, only Events emitted by the engine.